src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotRemoveBaseTypeOrInterface.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Microsoft.DotNet.ApiCompatibility.Extensions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    public class CannotRemoveBaseTypeOrInterface : Rule
+    {
+        public override void Initialize(RuleRunnerContext context)
+        {
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left == null || right == null)
+                return;
+
+            if (left.TypeKind != TypeKind.Interface && right.TypeKind != TypeKind.Interface)
+            {
+                // if left and right are not interfaces check base types
+                ValidateBaseTypeNotRemoved(left, right, leftName, rightName, differences);
+
+                if (Settings.StrictMode)
+                    ValidateBaseTypeNotRemoved(right, left, rightName, leftName, differences);
+            }
+
+            ValidateInterfaceNotRemoved(left, right, leftName, rightName, differences);
+
+            if (Settings.StrictMode)
+                ValidateInterfaceNotRemoved(right, left, rightName, leftName, differences);
+        }
+
+        private void ValidateBaseTypeNotRemoved(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            // Base type order matters in the hierarchy chain so therefore use a list.
+            List<ITypeSymbol> rightBaseTypes = right.GetAllBaseTypes().ToList();
+
+            int lastTypeIndex = 0;
+            foreach (ITypeSymbol leftBaseType in left.GetAllBaseTypes())
```

Comment:
Do we really need to iterate every base type, or can we just make sure that the immediate base type that left had is present in the heirarchy for right and then count on transitivity catching the rest of the heirarchy?

Reply:
I think we need to iterate all left base types if we want to report all removals that affect right in the chain, as this test captures:

Left: https://github.com/dotnet/sdk/blob/a6011df3ec4fa79535da9cb982f647456d1b1e66/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotRemoveBaseTypeOrInterfaceTests.cs#L231-L239
Right: https://github.com/dotnet/sdk/blob/a6011df3ec4fa79535da9cb982f647456d1b1e66/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotRemoveBaseTypeOrInterfaceTests.cs#L255-L265
Result: https://github.com/dotnet/sdk/blob/a6011df3ec4fa79535da9cb982f647456d1b1e66/src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/CannotRemoveBaseTypeOrInterfaceTests.cs#L297-#L299

However, we could safely assume that the only base type on left that we need to make sure exists in right is the first immediate base type in left. Reasons to assume that is safe, because if any base type is removed on the chain other than the first one on left, that will be caught when we compare the type which it's base type was removed.

So in the example above, the only difference we would get, it would be:

_SecondBase doesn't inherit from `ThirdBase` anymore._

Rather than getting the error of `ThirdBase` being remove on both, `First` and `SecondBase`.

Reply:
Using the diagnostics of the break is a good measurement of which is more correct.  Does one approach seem more accurate than the other based on the diagnostics raised?  I'm neutral on preference here, but raised this as a possibility in case it might simplify the code at all.